corehq/apps/hqadmin/management/commands/republish_doc_changes.py

```diff
@@ -17,62 +19,75 @@
 
-    def handle(self, domain, doc_ids_file, *args, **options):
-        document_records = _get_document_records(doc_ids_file)
+    def handle(self, stale_data_in_es_file, delimiter, *args, **options):
+        data_rows = _get_data_rows(stale_data_in_es_file, delimiter=delimiter)
+        document_records = _get_document_records(data_rows)
         form_records = []
         case_records = []
         for record in document_records:
-            if record.doc_type == CASE_DOC_TYPE:
+            if record.doc_type in CASE_DOC_TYPES:
                 case_records.append(record)
-            else:
-                assert record.doc_type == FORM_DOC_TYPE
+            elif record.doc_type in FORM_DOC_TYPES:
                 form_records.append(record)
+            else:
+                assert False, f'Bad doc type {record.doc_type} should have been caught already below.'
+        _publish_cases(case_records)
+        _publish_forms(form_records)
+
 
-        _publish_cases(domain, case_records)
-        _publish_forms(domain, form_records)
+def _get_data_rows(stale_data_in_es_file, delimiter):
+    with open(stale_data_in_es_file, 'r') as f:
+        for line in f.readlines():
+            data_row = DataRow(*line.rstrip('\n').split(delimiter))
+            # Skip the header row anywhere in the file.
+            # The "anywhere in the file" part is useful
+            # if you cat multiple stale_data_in_es_file files together.
+            if data_row != HEADER_ROW:
+                yield data_row
 
 
-def _get_document_records(doc_ids_file):
-    with open(doc_ids_file, 'r') as f:
-        lines = f.readlines()
-        for l in lines:
-            doc_id, doc_type, doc_subtype = [val.strip() for val in l.split(',')[0:3]]
-            if doc_type not in [CASE_DOC_TYPE, FORM_DOC_TYPE]:
-                raise CommandError(f"Found bad doc type {doc_type}. "
-                                   "Did you use the right command to create the data?")
-            yield DocumentRecord(doc_id, doc_type, doc_subtype)
+def _get_document_records(data_rows):
+    for data_row in data_rows:
+        doc_id, doc_type, doc_subtype, domain = \
+            data_row.doc_id, data_row.doc_type, data_row.doc_subtype, data_row.domain
+        if doc_type not in set.union(CASE_DOC_TYPES, FORM_DOC_TYPES):
+            raise CommandError(f"Found bad doc type {doc_type}. "
+                               "Did you use the right command to create the data?")
+        yield DocumentRecord(doc_id, doc_type, doc_subtype, domain)
 
 
-def _publish_cases(domain, case_records):
-    if should_use_sql_backend(domain):
-        _publish_cases_for_sql(domain, case_records)
-    else:
-        _publish_cases_for_couch(domain, [c.doc_id for c in case_records])
+def _publish_cases(case_records):
+    for domain, records in itertools.groupby(case_records, lambda r: r.domain):
+        if should_use_sql_backend(domain):
+            _publish_cases_for_sql(domain, records)
```

Comment:
This is the only blocker. All others are optional/suggestions.

Reply:
Thanks for catching this. I'm tempted to just throw away the second bit that relies on the second run through; if you are uploading the output of `stale_data_in_es` then the second bit will always be an empty list, and seems weird to be selectively supporting things that don't match the intended usage.

Reply:
@czue what do you think about dropping the section where we look up a case's case_type if it's not in the input file? https://github.com/dimagi/commcare-hq/blob/7923460997afedb252dfb4a35e2267665df5acff/corehq/apps/hqadmin/management/commands/republish_doc_changes.py#L116-L131. Looks like previously this command was made for dealing with just a list of IDs, and that's where this behavior comes from, but now if you're following the process the input file will already have this info. Also, other functionality has since been added, and it's likely that other things break the ability to import with just a list of ids. Thoughts?

Reply:
:+1:  that makes sense to me.